Fix bgpcfgd error of "Peer-group members must be all internal or all external" for some topos

byu343 opened the pull request:
…



<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/sonic-net/SONiC/blob/gh-pages/CONTRIBUTING.md

Please provide following information to help code review process a bit easier:
-->
### Description of PR
<!--
- Please include a summary of the change and which issue is fixed.
- Please also include relevant motivation and context. Where should reviewer start? background context?
- List any dependencies that are required for this change.
-->

Summary:
Fixes # (issue)

### Type of change

<!--
- Fill x for your type of change.
- e.g.
- [x] Bug fix
-->

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [ ] New Test case
    - [ ] Skipped for non-supported platforms
- [ ] Test case improvement


### Back port request
- [ ] 202012
- [ ] 202205
- [ ] 202305
- [ ] 202311
- [ ] 202405
- [ ] 202411

### Approach
#### What is the motivation for this PR?
This change is to fix the issue of bgpcfgd error of "Peer-group members must be all internal or all external" for topos t0-isolated-d16u16s2, t0-isolated-d96u32s2 and t0-isolated-d128u128s2.

#### How did you do it?
We followed the changes below to avoid existence of internal bgp by changing ASN of peer switches:
https://github.com/sonic-net/sonic-mgmt/blob/master/ansible/vars/topo_t0-isolated-d32u32s2.yml and PR https://github.com/sonic-net/sonic-mgmt/pull/17888
Then the routes to peer switches is added to the existing ecmp routes originally only to t1 switch. The added routes will break the assumption of existing testcases, so we have to change anounce_routes.py so that the routes to peer switches are not announced.

#### How did you verify/test it?
Tested that this will fix the bgpcfgd error of "Peer-group members must be all internal or all external"

#### Any platform specific information?

#### Supported testbed topology if it's a new test case?

### Documentation
<!--
(If it's a new feature, new test case)
Did you update documentation/Wiki relevant to your implementation?
Link to the wiki page?
-->

mssonicbld commented:
/azp run

StormLiangMS commented:
/azp run

mssonicbld commented:
/azp run

StormLiangMS commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

StormLiangMS commented:
hi @sdszhang could you help to take a look?

sdszhang commented on ansible/vars/topo_t0-isolated-d128u128s2.yml:
```diff
@@ -3152,7 +3152,7 @@ configuration:
     bgp:
-      asn: 65100
+      asn: 65101
       peers:
```
please update generate_topo.py as well.

byu343 commented on ansible/library/announce_routes.py:
```diff
@@ -1469,6 +1471,7 @@ def main():
 
     is_storage_backend = "backend" in topo_name
     tor_default_route = "t1-isolated" in topo_name
+    skip_peer_switches = topo_name in ['t0-isolated-d16u16s2', 't0-isolated-d96u32s2', 't0-isolated-d128u128s2']
```
Fixed

byu343 commented on ansible/library/announce_routes.py:
```diff
@@ -500,6 +500,8 @@ def fib_t0(topo, ptf_ip, no_default_route=False, action="announce"):
 
     vms = topo['topology']['VMs']
     for vm_name, vm in vms.items():
+        if 'P' in vm_name and skip_peer_switches:
```
Fixed

mssonicbld commented:
/azp run

StormLiangMS commented:
/azp run

sdszhang commented:
hi @byu343, can you try on latest 202412 and see if the issue is still there? We have updated the template used for PT0 in  https://github.com/sonic-net/sonic-mgmt/pull/18995, it should use a different peer-group now. ?

byu343 commented:
> hi @byu343, can you try on latest 202412 and see if the issue is still there? We have updated the template used for PT0 in #18995, it should use a different peer-group now. ?

I tested with reverting our PR entirely and including PR 18995, there seems still only PEER_V4 group. Randomly, it could be PT0 devices or T1 devices failed to be added to group for the error below. 

```
2025 Jun 19 23:09:46.986511 moby201 ERR bgp#bgpcfgd: command execution returned 13. Command: '['vtysh', '-f', '/tmp/tmp4a__ynln']', stdout: '[84|mgmtd] sending configuration#012[85|zebra] sending configuration#012[91|bgpd] sending configuration#012[101|staticd] sending configuration#01
2Waiting for children to finish applying config...#012% Peer-group members must be all internal or all external.#012% Peer-group members must be all internal or all external.#012% Peer-group members must be all internal or all external.#012% Peer-group members must be all internal or a
ll external.#012', stderr: '[101|staticd] done#012[85|zebra] done#012line 1225: Failure to communicate[13] to bgpd, line:     neighbor fc00::206 peer-group PEER_V6#012#012line 2341: Failure to communicate[13] to bgpd, line:     neighbor 10.0.1.1 peer-group PEER_V4#012#012line 2589: Fai
lure to communicate[13] to bgpd, line:     neighbor fc00::202 peer-group PEER_V6#012#012line 4697: Failure to communicate[13] to bgpd, line:     neighbor 10.0.1.3 peer-group PEER_V4#012#012[84|mgmtd] done#012[91|bgpd] Configuration file[/etc/frr/frr.conf] processing failure: 13#012'
2025 Jun 19 23:09:46.986533 moby201 ERR bgp#bgpcfgd: ConfigMgr::commit(): can't push configuration from file='/tmp/tmp4a__ynln', rc='13', stdout='[84|mgmtd] sending configuration#012[85|zebra] sending configuration#012[91|bgpd] sending configuration#012[101|staticd] sending configurati
on#012Waiting for children to finish applying config...#012% Peer-group members must be all internal or all external.#012% Peer-group members must be all internal or all external.#012% Peer-group members must be all internal or all external.#012% Peer-group members must be all internal
 or all external.#012', stderr='[101|staticd] done#012[85|zebra] done#012line 1225: Failure to communicate[13] to bgpd, line:     neighbor fc00::206 peer-group PEER_V6#012#012line 2341: Failure to communicate[13] to bgpd, line:     neighbor 10.0.1.1 peer-group PEER_V4#012#012line 2589:
 Failure to communicate[13] to bgpd, line:     neighbor fc00::202 peer-group PEER_V6#012#012line 4697: Failure to communicate[13] to bgpd, line:     neighbor 10.0.1.3 peer-group PEER_V4#012#012[84|mgmtd] done#012[91|bgpd] Configuration file[/etc/frr/frr.conf] processing failure: 13#012
```

The initial purpose of this PR is avoid that error, implemented by changing ASN of PT0 device in the topo files.
After it, another issue came up which is the PT0 peers added to the multiple patch routes, so we tried to avoid that by changing ansible/library/announce_routes.py

mssonicbld commented:
/azp run

StormLiangMS commented:
/azp run

sdszhang commented on ansible/vars/topo_t0-isolated-d128u128s2.yml:
```diff
@@ -733,13 +733,13 @@ configuration:
     - common
     bgp:
-      asn: 64600
+      asn: 64602
```
the PT0 ASN number is fixed in https://github.com/sonic-net/sonic-mgmt/pull/19526
